File: apps/android/app/src/main/java/com/mxclaw/app/network/WebSocketManager.kt
package com.mxclaw.app.network

import com.mxclaw.app.model.ChatMessage
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.TimeUnit

class WebSocketManager {

    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private var webSocket: WebSocket? = null

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _messages = MutableStateFlow<List<ChatMessage>>(emptyList())
    val messages: StateFlow<List<ChatMessage>> = _messages.asStateFlow()

    private val _pairedDeviceId = MutableStateFlow<String?>(null)
    val pairedDeviceId: StateFlow<String?> = _pairedDeviceId.asStateFlow()

    var serverUrl: String
        get() = prefsGet("server_url", "ws://localhost:18700")
        set(value) = prefsSet("server_url", value)

    fun connect() {
        disconnect()
        val url = serverUrl
        val request = Request.Builder()
            .url(url)
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _isConnected.value = true
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _isConnected.value = false
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _isConnected.value = false
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "Client closing")
        webSocket = null
        _isConnected.value = false
    }

    fun sendChat(text: String) {
        val uuid = UUID.randomUUID().toString()
        val localMsg = ChatMessage(
            id = uuid,
            role = "user",
            content = text,
            timestamp = null
        )
        val current = _messages.value.toMutableList()
        current.add(localMsg)
        _messages.value = current

        val payload = JSONObject().apply {
            put("type", "chat:send")
            put("message", text)
            put("id", uuid)
        }
        webSocket?.send(payload.toString())
    }

    fun sendPairCode(code: String) {
        val payload = JSONObject().apply {
            put("type", "pair:request")
            put("code", code)
        }
        webSocket?.send(payload.toString())
    }

    private fun handleMessage(text: String) {
        try {
            val json = JSONObject(text)
            when (json.optString("type")) {
                "pair:success" -> {
                    val deviceId = json.optString("deviceId")
                    _pairedDeviceId.value = deviceId
                }
                else -> {
                    val msg = ChatMessage.fromJson(json)
                    if (msg.id.isNotBlank()) {
                        val current = _messages.value.toMutableList()
                        current.add(msg)
                        _messages.value = current
                    }
                }
            }
        } catch (_: Exception) {
        }
    }

    private fun prefsGet(key: String, default: String): String {
        // Simplified - in real app use SharedPreferences
        return default
    }

    private fun prefsSet(key: String, value: String) {
        // Simplified - in real app use SharedPreferences
    }
}

File: apps/ios/MxClaw/Sources/WebSocketManager.swift
import Foundation
import Combine

class WebSocketManager: ObservableObject {
    @Published var isConnected = false
    @Published var messages: [ChatMessage] = []
    @Published var pairedDeviceId: String?

    private var webSocketTask: URLSessionWebSocketTask?
    private var session: URLSession?

    var serverURL: String {
        get { UserDefaults.standard.string(forKey: "serverURL") ?? "ws://localhost:18700" }
        set { UserDefaults.standard.set(newValue, forKey: "serverURL") }
    }

    func connect() {
        disconnect()
        guard let url = URL(string: serverURL) else { return }
        session = URLSession(configuration: .default, delegate: nil, delegateQueue: .main)
        webSocketTask = session?.webSocketTask(with: url)
        webSocketTask?.resume()
        isConnected = true
        receiveMessage()
    }

    func disconnect() {
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        session?.invalidateAndCancel()
        session = nil
        isConnected = false
    }

    func sendChat(_ text: String) {
        let uuid = UUID().uuidString
        let localMsg = ChatMessage(
            id: uuid,
            role: "user",
            content: text,
            timestamp: ISO8601DateFormatter().string(from: Date())
        )
        DispatchQueue.main.async {
            self.messages.append(localMsg)
        }
        let payload: [String: Any] = [
            "type": "chat:send",
            "message": text,
            "id": uuid
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload) else { return }
        webSocketTask?.send(.data(data)) { error in
            if let error = error {
                print("Send error: \(error)")
            }
        }
    }

    func sendPairCode(_ code: String) {
        let payload: [String: Any] = [
            "type": "pair:request",
            "code": code
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload) else { return }
        webSocketTask?.send(.data(data)) { error in
            if let error = error {
                print("Pair error: \(error)")
            }
        }
    }

    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(let message):
                switch message {
                case .data(let data):
                    self.handleData(data)
                case .string(let string):
                    if let data = string.data(using: .utf8) {
                        self.handleData(data)
                    }
                @unknown default:
                    break
                }
                self.receiveMessage()
            case .failure(let error):
                print("Receive error: \(error)")
                DispatchQueue.main.async {
                    self.isConnected = false
                }
            }
        }
    }

    private func handleData(_ data: Data) {
        if let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let type = dict["type"] as? String {
                switch type {
                case "pair:success":
                    if let deviceId = dict["deviceId"] as? String {
                        DispatchQueue.main.async {
                            self.pairedDeviceId = deviceId
                        }
                    }
                case "chat:message":
                    if let msg = try? JSONDecoder().decode(ChatMessage.self, from: data) {
                        DispatchQueue.main.async {
                            self.messages.append(msg)
                        }
                    }
                default:
                    if let msg = try? JSONDecoder().decode(ChatMessage.self, from: data) {
                        DispatchQueue.main.async {
                            self.messages.append(msg)
                        }
                    }
                }
            }
        }
    }
}

struct ChatMessage: Identifiable, Decodable {
    let id: String
    let role: String
    let content: String
    let timestamp: String?

    enum CodingKeys: String, CodingKey {
        case id, role, content, timestamp
    }
}
